ao-tools/ao-eeprom/ao-eeprom.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-EEPROM 1 "ao-eeprom" ""
.SH NAME
ao-eeprom \- Fetch eeprom contents from TeleMetrum device
.SH SYNOPSIS
.B "ao-eeprom"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMetrum:2
.br
TeleMetrum
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.SH DESCRIPTION
.I ao-eeprom
downloads the eeprom contents from a connected TeleMetrum device.
.SH USAGE
.I ao-eeprom
connects to the specified target device and dumps each block of the
eeprom to stdout in hexidecimal bytes.
.SH AUTHOR
Keith Packard


ChangeLog
commit a5e94aa0677070a051714443cf7fd7e2b5e90269
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:59:47 2009 -0600

    need a run-time dependency to pull in the cairo module

commit 97acef95cc9843998963921459fdd71dd7eaa6b4
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:48:23 2009 -0600

    update changelogs for Debian build

commit 3f95a5abbf8ada70328ced45fbb2781ed1cb3d29
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:48:10 2009 -0600

    more build deps for plotting lib

commit 9d7e96e323d652de08b2f2a73e0eb3c321080185
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:47:56 2009 -0600

    update changelogs for Debian build

commit ae4e131b61244e06020b82919e31e05dd7dba88f
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:46:39 2009 -0600

    update changelogs for Debian build

commit 37e6c9a492a1d51373bf9333fb3172e0c377720f
Merge: d256f82 2e6686b
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:46:10 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 2e6686b1e183c66188ea447b8a54e4c29402443b
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 16:45:47 2009 -0700

    Use plplotd instead of plplotd-gnome2

commit d256f8204e9fce53ae4309562bb4c0cde1fae43e
Merge: 0fc344d 32d3536
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 17:34:08 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 32d3536706324808df6cd02248a236302b831571
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 16:24:35 2009 -0700

    Add plots to ao-postflight using the plplot library
    
    It's not perfect, but it generates .svg plot output.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0fc344dfc031a8b1eef7cc40efb1d5ba7782269d
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:15:57 2009 -0600

    update changelogs for Debian build

commit 4b0de757874c0ecaf38e3dfd3beefc398150e3d5
Merge: 773c4ff d0eac98
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:15:53 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit d0eac989b1ffc8ae30ba12da403eb4bf1ad42d6b
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 13:15:10 2009 -0700

    Don't look at NULL strings (summary_name)
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 73f4a57239f770aff603b961169c0e2cfe2c276b
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 13:08:54 2009 -0700

    Use pressure speed for drogue and beyond states. Fix differentiation time.
    
    Drogue state should always use pressure speeds.
    
    Differentiation code was using centi-seconds instead of seconds.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 773c4ffbc1d2e02eb02cfa543a077a408986da30
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:05:55 2009 -0600

    update changelogs for Debian build

commit 45ede4a4b203ef9da5bf05c49cb9c5a2e6382ec5
Merge: 45e2938 e35e485
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:05:51 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit e35e485ffe6b26034788ab295121bc2693b7eec1
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 13:04:31 2009 -0700

    Initialize summary_name and detail_name so stuff appears on stdout.
    
    Uninitialized variables lead to mysterious results.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 45e2938121411d1fc9b3aec3fdeaaeb3c90db5ed
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:02:14 2009 -0600

    update changelogs for Debian build

commit d42ebf0661ecf15455e5051de1e16ae66f8dd857
Merge: 384dbe9 7a19aac
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 14:02:09 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 7a19aac5e881e635962a64fff73027ca2143b96f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 6 12:51:48 2009 -0700

    Add DSP code to filter data, allowing for integration/differentiation
    
    This adds the computation of speed from both accelerometer and
    barometer measurements and then presents a periodic flight profile
    using filtered data as a detailed flight record.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 384dbe9fc7fa8e4e5dceec5e150f0f1d3383bbdc
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 10:40:06 2009 -0600

    update changelogs for Debian build

commit 35c54b3a278fa9bc2bc7f4b5ee04866697c93ba0
Merge: 4f8eff7 6d018ab
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Sep 6 10:39:23 2009 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 6d018ab933832e2d80bb1564c339d9fb18b57be2
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Sep 5 22:45:49 2009 -0700

    Handle vageries of .telem files in ao-postflight
    
    Telem files have multiple entries of the same state, and sometimes
    long gaps between recordings. Deal with this as best as possible.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c46e832b28820d7c5be4efaacbbd7c0607927fe5
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Sep 5 22:03:31 2009 -0700

    Add simple post-flight analysis tool (ao-postflight)
    
    This tool reads either an eeprom or telem log file and displays some
    rudimentary data (max accel/alt for each flight stage).
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 26f56b51bd11aa91f1d77b81827b49c28cb6ec5f
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Sep 5 00:29:26 2009 -0700

    Add ao-dumplog to capture flight log from command line
    
    This duplicates the functionality of the flight log stuf in ao-view,
    except from the command line where it belongs.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 73adae3661160d410dcc802873b530d255c210e5
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 4 15:30:22 2009 -0700

    Add --device/-D support to the command line tools and manuals

ao-tools/ao-eeprom/ao-eeprom.c
/*
 * Copyright © 2009 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include "cc-usb.h"
#include "cc.h"

#define NUM_BLOCK	512

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>\n", program);
	exit(1);
}

int
main (int argc, char **argv)
{
	struct cc_usb	*cc;
	int		block;
	uint8_t		bytes[2 * 32 * (2 + 8)];
	uint8_t		*b;
	int		i, j;
	uint32_t	addr;
	char		*tty = NULL;
	char		*device = NULL;
	int		c;

	while ((c = getopt_long(argc, argv, "T:D:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "TeleMetrum");
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";
	cc = cc_usb_open(tty);
	if (!cc)
		exit(1);
	for (block = 0; block < NUM_BLOCK; block += 2) {
		cc_queue_read(cc, bytes, sizeof (bytes));
		cc_usb_printf(cc, "e %x\ne %x\n", block, block + 1);
		cc_usb_sync(cc);
		for (i = 0; i < 32 * 2; i++) {
			b = bytes + (i * 10);
			addr = block * 256 + i * 8;
			printf ("%06x", addr);
			for (j = 0; j < 8; j++) {
				printf (" %02x", b[j+2]);
			}
			printf ("\n");
		}
	}
	cc_usb_close(cc);
	exit (0);
}
